### PR TITLE
HPCC-25465 Message collator insert method can be n^2

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: roxie/udplib/udpmsgpk.cpp [hunk 1]
```diff
@@ -250,6 +292,12 @@ class PackageSequencer : public CInterface, implements IInterface
         return headerSize;
     }
 
+#ifdef _DEBUG
+    void dump()
+    {
+        DBGLOG("Contains %d packets, lastSeq = %d", numPackets, maxSeqSeen);
```
Comment (ghalliday) [inline roxie/udplib/udpmsgpk.cpp:298]: really should be %u, instead of %d

**Review by ghalliday — COMMENTED** [review]
A couple of minor comments, but the code looks good.  II would get Mark to double check,and if he is happy it would be reasonable to include in 6.4.x

**Review by ghalliday — APPROVED** [review]

File: roxie/udplib/udpmsgpk.cpp [hunk 2]
```diff
@@ -122,54 +130,85 @@ class PackageSequencer : public CInterface, implements IInterface
...
-                // discard duplicated incoming packet - should be uncommon unless we requested a resend for a packet already in flight
-                if (checkTraceLevel(TRACE_MSGPACK, 5))
-                    DBGLOG("UdpCollator: Discarding duplicate incoming packet");
-                dataBuff->Release();
-                return false;
-            }
-            finger = lastContiguousPacket->msgNext;
-            prev = lastContiguousPacket;
+            finger = nullptr;
+            prev = tail;
```
Comment (richardkchapman) [inline roxie/udplib/udpmsgpk.cpp:151]: May need tail->msgNext = dataBuffer here...

File: roxie/udplib/udpmsgpk.cpp [hunk 3]
```diff
@@ -178,8 +217,11 @@ class PackageSequencer : public CInterface, implements IInterface
             prev->msgNext = dataBuff;
         }
         else
-            firstPacket = dataBuff;
+            firstPacket = tail = dataBuff;
```
Comment (richardkchapman) [inline roxie/udplib/udpmsgpk.cpp:220]: This incorrectly overwrites tail if packet 0 is not the first to be seen

File: roxie/udplib/udpmsgpk.cpp [hunk 4]
```diff
@@ -122,54 +130,86 @@ class PackageSequencer : public CInterface, implements IInterface
...
+                }
+                finger = lastContiguousPacket->msgNext;
+                prev = lastContiguousPacket;
             }
-            else if (pktHdr->pktSeq < oldHdr->pktSeq)
+            else
             {
-                break;
+                finger = firstPacket;
+                prev = NULL;
```
Comment (jakesmith) [inline roxie/udplib/udpmsgpk.cpp:175]: super trivia: could be nullptr

File: roxie/udplib/udpmsgpk.cpp [hunk 5]
```diff
@@ -122,54 +130,86 @@ class PackageSequencer : public CInterface, implements IInterface
...
+                        UdpPacketHeader *oldHdr = (UdpPacketHeader*) lastContiguousPacket->data;
+                        DBGLOG("lastContiguousPacket is at %u , last packet seen is %u", oldHdr->pktSeq & UDP_PACKET_SEQUENCE_MASK, pktHdr->pktSeq & UDP_PACKET_SEQUENCE_MASK);
+                    }
+                    else
+                        DBGLOG("lastContiguousPacket is NULL , last packet seen is %u", pktHdr->pktSeq & UDP_PACKET_SEQUENCE_MASK);
+                    scans = 0;
+                }
+    #endif
+                UdpPacketHeader *oldHdr = (UdpPacketHeader*) finger->data;
+                if (pktHdr->pktSeq == oldHdr->pktSeq)
```
Comment (jakesmith) [inline roxie/udplib/udpmsgpk.cpp:196]: so finger is walking through from lastContiguousPacket through msgNext,
is this possible, not already rejected at line 161? Or are the chained pktSeq values expected to be > than the lastContiguousPacket->pkgSeq ?

**Review by jakesmith — COMMENTED** [review]
@richardkchapman - 1 question that probably stems from lack of understanding of impl.

File: roxie/udplib/udpmsgpk.cpp (hunk 4 above)
Comment (richardkchapman) [reply roxie/udplib/udpmsgpk.cpp:175]: Only the indentation has changed here

File: roxie/udplib/udpmsgpk.cpp (hunk 5 above)
Comment (richardkchapman) [reply roxie/udplib/udpmsgpk.cpp:196]: The chain is always in order. Newly arriving packets are added via insertion sort. We make this more efficient by (a) tracking the leading already-complete portion and not trying to insert in there and (b) the new code - checking if the addition is at the end - this is true 99.9% of the time, at least if packets arrive in the same order that they were sent.

If packets are NOT received in the same order than they are sent then this change will not really help (but won't make anything any worse either).

Comment (jakesmith) [reply roxie/udplib/udpmsgpk.cpp:196]: makes sense now. Looks fine.

**Review by jakesmith — APPROVED** [review]

**Review by mckellyln — APPROVED** [review]
Looks good, just a comment about adding scan tracing.

**Pull request merged**